### PR TITLE
Fix demo content not working in WordPress 5.3

### LINKED ISSUE #5443: Demo content not working in WordPress 5.3
## Bug Description

Demo content not working in WordPress 5.3. 

## Screenshots

Screenshot 2020-11-25 at 13 21 41

## Additional Context

- Plugin Version: 1.1.1
- WordPress Version: 5.3
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Demo content not working in WordPress 5.3.

## Relevant Technical Choices

Demo content does not work as preloading in WordPress 5.3 is broken. The cache key was not generated correctly. It seems to be fixed in WordPress 5.4 and https://github.com/WordPress/gutenberg/pull/18724

This PR, makes sure that GET value is passed down to the API request. 

This is 

## Testing Instructions

---

Fixes #5443

### CODE REVIEW

File: includes/REST_API/Stories_Controller.php [hunk 1]
```diff
@@ -388,6 +389,12 @@ public function get_collection_params() {
 			'default'     => false,
 		];
 
+		$query_params['web_stories_demo'] = [
+			'description' => __( 'Load demo data.', 'web-stories' ),
+			'type'        => 'string',
```
Comment (swissspidy) [inline includes/REST_API/Stories_Controller.php]: Why not a bool?

**Comment by swissspidy** [comment]
Great find & nice workaround 👍 

The only thing I don't find ideal is how `getStoryById` now has to be aware of whether this is a demo request or not.

I think this should be a separate callback instead.

So we could do:

```js
// In apiProvider.js

const getDemoStory = useCallback(
  (storyId) => {
    const path = addQueryArgs(`${stories}${storyId}/`, {
      context: 'edit',
      _embed: 'wp:featuredmedia,author',
      web_stories_demo: 1
    });

    return apiFetch({ path });
  },
  [stories]
);

// In storyProvider.js

const { useDemoContent } = useConfig();

// ...

// This effect loads and initialises the story on first load (when there's no pages).
const shouldLoad = pages.length === 0;
useLoadStory({ restore, shouldLoad, storyId, useDemoContent });

// In useLoadStory.js

// ...

const callback = useDemoContent ? getDemoStory : getStoryById

useEffect(() => {
  if (storyId && shouldLoad) {
    callback(storyId).then((post) => {
	// ...
```

cc @barklund for thoughts

File: includes/REST_API/Stories_Controller.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/REST_API/Stories_Controller.php]: I used a string here, as it was a string before and it was a nice drop in replacement. Willing to change it, if it makes sense.

File: assets/src/edit-story/app/api/apiProvider.js [hunk 2]
```diff
@@ -47,6 +47,20 @@ function APIProvider({ children }) {
       const path = addQueryArgs(`${stories}${storyId}/`, {
         context: 'edit',
         _embed: 'wp:featuredmedia,author',
+        web_stories_demo: '0',
```
Comment (swissspidy) [inline assets/src/edit-story/app/api/apiProvider.js]: Why is this still needed here?

File: includes/REST_API/Stories_Controller.php (hunk 1 above)
Comment (swissspidy) [reply includes/REST_API/Stories_Controller.php]: It's a boolean flag after all, so we should use the correct type here

File: assets/src/edit-story/app/api/apiProvider.js (hunk 2 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/api/apiProvider.js]: The param is passed in prefetching.

File: includes/REST_API/Stories_Controller.php [hunk 3]
```diff
@@ -62,8 +64,7 @@ class Stories_Controller extends Stories_Base_Controller {
 	public function prepare_item_for_response( $post, $request ) {
 		$context = ! empty( $request['context'] ) ? $request['context'] : 'view';
 
-		// $_GET param is available when the response iss preloaded in edit-story.php
-		if ( isset( $_GET['web-stories-demo'] ) && 'edit' === $context && 'auto-draft' === $post->post_status ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( isset( $request['web_stories_demo'] ) && '1' === $request['web_stories_demo'] && 'edit' === $context && 'auto-draft' === $post->post_status ) {
```
Comment (swissspidy) [inline includes/REST_API/Stories_Controller.php]:
```suggestion
		if ( ! empty( $request['web_stories_demo'] ) && 'auto-draft' === $post->post_status ) {
```

File: includes/Story_Post_Type.php [hunk 4]
```diff
@@ -513,6 +515,7 @@ public function get_editor_settings() {
 				'storyId'          => $story_id,
 				'previewLink'      => get_preview_post_link( $story_id, $preview_query_args ),
 				'maxUpload'        => $max_upload_size,
+				'useDemoContent'   => $demo,
```
Comment (barklund) [inline includes/Story_Post_Type.php]: Things named `useFoo` in React are generally expected to be hooks and breaking this assumption can confuse devs (and tooling). Booleans really should be named `isFoo`, `hasFoo`, or `shouldFoo`. So I recommend changing it to `shouldUseDemoContent` or simply `isDemo` throughout the application.

**Review by barklund — COMMENTED** [review]
LGTM except for naming.

File: includes/REST_API/Stories_Controller.php (hunk 1 above)
Comment (swissspidy) [reply includes/REST_API/Stories_Controller.php]: If we can make it a bool then I am happy to approve 👍

**Review by swissspidy — APPROVED** [review]
LGTM pending tests passing

File: includes/REST_API/Stories_Controller.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/REST_API/Stories_Controller.php]: Changed in https://github.com/google/web-stories-wp/pull/5507/commits/a12720f45b99157d1675ef5bfdba28fed0ef4f0a

File: includes/Story_Post_Type.php (hunk 4 above)
Comment (spacedmonkey) [reply includes/Story_Post_Type.php]: Fixed in https://github.com/google/web-stories-wp/pull/5507/commits/48833436b29c22caea967b6c7870061bbb1f769e

File: includes/Story_Post_Type.php [hunk 5]
```diff
@@ -523,7 +523,7 @@ public function get_editor_settings() {
 			'preview_nonce' => wp_create_nonce( 'post_preview_' . $story_id ),
 		];
 
-		$demo = isset( $_GET['web-stories-demo'] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		$is_demo = ( isset( $_GET['web-stories-demo'] ) && boolval( $_GET['web-stories-demo'] ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
Comment (swissspidy) [inline includes/Story_Post_Type.php]: Should use `(bool) for type case, not `boolval`

Comment (spacedmonkey) [reply includes/Story_Post_Type.php]: Fixed in https://github.com/google/web-stories-wp/pull/5507/commits/d279dbf3036dbfc8045b3a66b8563a689c67e32c

File: includes/REST_API/Stories_Controller.php [hunk 6]
```diff
@@ -62,8 +64,7 @@ class Stories_Controller extends Stories_Base_Controller {
 	public function prepare_item_for_response( $post, $request ) {
 		$context = ! empty( $request['context'] ) ? $request['context'] : 'view';
 
-		// $_GET param is available when the response iss preloaded in edit-story.php
-		if ( isset( $_GET['web-stories-demo'] ) && 'edit' === $context && 'auto-draft' === $post->post_status ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( wp_validate_boolean( $request['web_stories_demo'] ) && 'auto-draft' === $post->post_status ) {
```
Comment (swissspidy) [inline includes/REST_API/Stories_Controller.php:67]: Totally forgot about `wp_validate_boolean`. Neat 👍

**Pull request merged**